include/deal.II/lac/vector_space_vector.h

```diff
@@ -1,304 +0,0 @@
-     * @code
-     * this->add(a, V);
-     * return_value = *this * W;
-     * @endcode
-     *
-     * The reason this function exists is that this operation involves less
-     * memory transfer than calling the two functions separately. This method
-     * only needs to load three vectors, @p this, @p V, @p W, whereas calling
-     * separate methods means to load the calling vector @p this twice. Since
-     * most vector operations are memory transfer limited, this reduces the
-     * time by 25\% (or 50\% if @p W equals @p this).
-     *
-     * For complex-valued vectors, the scalar product in the second step is
-     * implemented as
-     * $\left<v,w\right>=\sum_i v_i \bar{w_i}$.
-     */
-    virtual Number
-    add_and_dot(const Number                     a,
-                const VectorSpaceVector<Number> &V,
-                const VectorSpaceVector<Number> &W) = 0;
-
-    /**
-     * This function does nothing and only exists for backward compatibility.
-     */
-    virtual void compress(VectorOperation::values)
-    {}
-
-    /**
-     * Return the global size of the vector, equal to the sum of the number of
-     * locally owned indices among all processors.
-     */
-    virtual size_type
-    size() const = 0;
-
-    /**
-     * Return an index set that describes which elements of this vector are
-     * owned by the current processor. As a consequence, the index sets
-     * returned on different processors if this is a distributed vector will
-     * form disjoint sets that add up to the complete index set. Obviously, if
-     * a vector is created on only one processor, then the result would
-     * satisfy
-     * @code
-     *  vec.locally_owned_elements() == complete_index_set(vec.size())
-     * @endcode
-     */
-    virtual dealii::IndexSet
-    locally_owned_elements() const = 0;
-
-    /**
-     * Print the vector to the output stream @p out.
-     */
-    virtual void
-    print(std::ostream &     out,
-          const unsigned int precision  = 3,
-          const bool         scientific = true,
-          const bool         across     = true) const = 0;
-
-    /**
-     * Return the memory consumption of this class in bytes.
-     */
-    virtual std::size_t
-    memory_consumption() const = 0;
-
-    /**
-     * Destructor. Declared as virtual so that inheriting classes (which may
-     * manage their own memory) are destroyed correctly.
-     */
-    virtual ~VectorSpaceVector() = default;
-  };
-  /** @} */
-} // namespace LinearAlgebra
-
-// ---------------------------- Free functions --------------------------
-
-namespace LinearAlgebra
-{
-  /**
-   * Shift all entries of the vector by a constant factor so that the mean
-   * value of the vector becomes zero.
-   */
-  template <typename Number>
-  void
-  set_zero_mean_value(VectorSpaceVector<Number> &vector)
-  {
-    vector.add(-vector.mean_value());
-  }
-} // namespace LinearAlgebra
```

Comment:
This is something we should move somewhere else. This change will break compilation of https://github.com/exadg/exadg for example. We should move this function somewhere else with `template <typename VectorType>` arguments, the question is where we want it to be to minimize backwards compatibility break.

Reply:
Good point. I could just put that back into this file as the only thing that's in there. Would that be acceptable?

Reply:
Do we have another file which is included by all vector classes? I think many use cases did just include the derived vector class, e.g. `deal.II/lac/la_parallel_vector.h` in our case of ExaDG, and got the declaration in `vector_space_vector.h` on the way. I'm not sure where it fits best, but the three options are `read_vector.h`, `vector_type_traits.h` and `vector_operation.h`, none of which uses the `LinearAlgebra` namespace. We could also set a fourth file `vector_templace_functions.h` or similar that all our vector classes include.

Reply:
`VectorTools::subtract_mean_value()` does the same thing:

https://github.com/dealii/dealii/blob/de6a2e7a6f4ffdc91b780e67b5e468f921f5a27c/include/deal.II/numerics/vector_tools_mean_value.templates.h#L90-L98

We could add the header back in but deprecate the function in favor of the `VectorTools` version.

Reply:
I agree.